main.py

```diff
@@ -333,22 +349,63 @@ def read_simulation_results(simulation_xml):
     last_X = last_rows[last_rows["population"] == "X"]["value"].values[0]
     last_Psi = last_rows[last_rows["population"] == "Psi"]["value"].values[0]
     last_Mu = last_rows[last_rows["population"] == "Mu"]["value"].values[0]
-    return {
+    sim_result_structure = {
         "tree": tree,
         "tree_height": max(tree.depths().values()),
         "present": last_psi_time,
         "present_prevalence": last_X,
-        "present_cumulative": last_Psi + last_Mu + last_X,
+        "present_cumulative": last_Psi + last_Mu + last_X
     }
+    
+    # ADDED 21-8: optionally report rzero, prevalence, cumul.
+    # infections, rzero over time here as a numpy recarray
+    if REPORT_TEMPORAL_DATA:
+    
+        meas_times = np.sort(np.random.uniform(low=0.0, high=sim_result_structure["present"],
+                                       size=NUM_TEMP_MEASUREMENTS))
+        r0_change_times = pd.Series(params["r0"]["change_times"])
+        
+        temp_data_headers = ["measurement-times", "prevalence", "cumulative", "reproductive-number"]
+        temp_data = [None for _ in range(NUM_TEMP_MEASUREMENTS)]
+        
+        for time_ind in range(NUM_TEMP_MEASUREMENTS):
+        
+            this_meas_time = meas_times[time_ind]
+        
+            most_recent_obs_time = traj_df[traj_df["t"] >= this_meas_time]["t"].min()
```

Comment:
Since the functions are piece-wise continuous with left limits, I think this should be

`most_recent_change_time = traj_df[traj_df["t"] >= this_meas_time]["t"].max()` 

so that it picks up the time the state changed to its current value. Am I missing something about the method you have used here?

Reply:
Yes, that's a good point, not sure what I did there? But will correct this. I think it needs to be

<code>most_recent_change_time = traj_df[traj_df["t"] <= this_meas_time]["t"].max()<\code>

i.e. the most recent change time prior to the requested measurement time. This is what I've implemented in the new version.